File: kms-frost/src/curve.rs
//! Signing scheme identifier shared across KMS modules.
//!
//! `Curve` identifies the signing scheme (not just the curve) for keygen,
//! sign, and storage. `Secp256k1` and `Ed25519` use FROST Schnorr.
//! `EcdsaSecp256k1` uses threshold ECDSA (DJNPO20) over secp256k1,
//! producing standard ECDSA signatures compatible with EVM ecrecover.

use serde::{Deserialize, Serialize};

/// Supported signing schemes.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum Curve {
    /// FROST Schnorr over secp256k1.
    Secp256k1,
    /// FROST Schnorr over Ed25519.
    Ed25519,
    /// Threshold ECDSA over secp256k1 (DJNPO20 robust protocol).
    /// Produces standard ECDSA signatures (r, s, v) compatible with ecrecover.
    /// Uses the same DKG/reshare as Secp256k1 (FROST keygen), but a different
    /// 4-round signing protocol.
    EcdsaSecp256k1,
}

impl Curve {
    /// Wire name used in CBOR params and API requests.
    pub fn as_str(&self) -> &'static str {
        match self {
            Curve::Secp256k1 => "secp256k1",
            Curve::Ed25519 => "ed25519",
            Curve::EcdsaSecp256k1 => "ecdsa_secp256k1",
        }
    }

    /// Single-byte prefix for storage keys. Ensures same key_id with
    /// different schemes maps to different storage entries.
    pub fn storage_prefix(&self) -> u8 {
        match self {
            Curve::Secp256k1 => 0x01,
            Curve::Ed25519 => 0x02,
            Curve::EcdsaSecp256k1 => 0x03,
        }
    }

    /// Whether this scheme uses secp256k1 keygen (FROST DKG).
    /// Both Secp256k1 (FROST Schnorr) and EcdsaSecp256k1 share the same
    /// keygen — only the signing protocol differs.
    pub fn uses_secp256k1_keygen(&self) -> bool {
        matches!(self, Curve::Secp256k1 | Curve::EcdsaSecp256k1)
    }
}

impl Curve {
    /// Map a frost_core Ciphersuite type to its Curve value.
    /// Used in generic code where C is known at compile time.
    pub fn from_ciphersuite<C: frost_core::Ciphersuite>() -> Self {
        // Compare group key serialization size: 33 = secp256k1, 32 = ed25519.
        // This is a compile-time-resolvable dispatch since the Serialization
        // type is fixed per Ciphersuite.
        let id = C::ID;
        if id.starts_with("FROST-secp256k1") {
            Curve::Secp256k1
        } else if id.starts_with("FROST-Ed25519") || id.starts_with("FROST-ED25519") {
            Curve::Ed25519
        } else {
            panic!("unknown ciphersuite ID: {id}");
        }
    }
}

impl std::fmt::Display for Curve {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.write_str(self.as_str())
    }
}

File: node/curve.go
package node

// Curve identifies a signing scheme for keygen, sign, and storage.
type Curve string

const (
	CurveSecp256k1      Curve = "secp256k1"
	CurveEd25519        Curve = "ed25519"
	CurveEcdsaSecp256k1 Curve = "ecdsa_secp256k1"
)

// Valid returns true if c is a recognized curve.
func (c Curve) Valid() bool {
	return c == CurveSecp256k1 || c == CurveEd25519 || c == CurveEcdsaSecp256k1
}

// String returns the wire name (used in CBOR params and API responses).
func (c Curve) String() string {
	return string(c)
}

// StoragePrefix returns the single-byte prefix used in KMS sled storage keys.
// Matches kms-tss/src/curve.rs Curve::storage_prefix().
func (c Curve) StoragePrefix() byte {
	switch c {
	case CurveSecp256k1:
		return 0x01
	case CurveEd25519:
		return 0x02
	case CurveEcdsaSecp256k1:
		return 0x03
	default:
		return 0x00
	}
}

// CurveFromStoragePrefix returns the Curve for a storage prefix byte.
func CurveFromStoragePrefix(b byte) (Curve, bool) {
	switch b {
	case 0x01:
		return CurveSecp256k1, true
	case 0x02:
		return CurveEd25519, true
	case 0x03:
		return CurveEcdsaSecp256k1, true
	default:
		return "", false
	}
}

// IsSecp256k1 returns true if this scheme uses secp256k1 keygen.
// Both Secp256k1 (FROST Schnorr) and EcdsaSecp256k1 share the same DKG.
func (c Curve) IsSecp256k1() bool {
	return c == CurveSecp256k1 || c == CurveEcdsaSecp256k1
}

// KeyEntry identifies a stored key by its ID and curve.
type KeyEntry struct {
	KeyID string
	Curve Curve
}
